src/ipynb.rs

```diff
@@ -192,101 +206,463 @@ fn notebook_path(env: &PyEnv, path: &PathStr) -> Result<PathBuf, NotebookToPytho
+    __aqora__script = compile(f.read(), __aqora__path, 'exec', flags=ast.PyCF_ALLOW_TOP_LEVEL_AWAIT)
+
+async def __aqora__(*__aqora__args, **__aqora__kwargs):
+    globals().update(locals())
+    coroutine = eval(__aqora__script, globals())
+    if coroutine is not None:
+        await coroutine
+    if 'output' in globals():
+        return globals()['output']
+    else:
+        raise NameError("No 'output' variable found in the notebook")
+"#,
+            name = self.generated_name
+        )
+    }
+
+    fn module_function_def(&self) -> String {
+        format!(
+            r#"
+spec_{name} = importlib.util.spec_from_file_location(
+    '{module_name}',
+    dir_path / 'generated' / '{name}.py')
+module_{name} = importlib.util.module_from_spec(spec_{name})
+sys.modules['{module_name}'] = module_{name}
+spec_{name}.loader.exec_module(module_{name})
+
+async def {func}(*__aqora__args, **__aqora__kwargs):
+    return await module_{name}.__aqora__(*__aqora__args, **__aqora__kwargs)
+"#,
+            func = self.function_name(),
+            module_name = self.path,
+            name = self.generated_name
+        )
+    }
+}
+
+fn get_meta(env: &PyEnv, path: &PathStr) -> Result<NotebookMeta, NotebookToPythonFunctionError> {
+    let notebook_path = notebook_path(env, path)?;
+    let notebook_path = notebook_path.canonicalize().map_err(|e| {
+        NotebookToPythonFunctionError::Read(
+            notebook_path.clone(),
+            std::io::Error::new(
+                std::io::ErrorKind::NotFound,
+                format!("Could not find notebook: {}", e),
+            ),
+        )
+    })?;
+    let generated_name = base32::encode(
+        base32::Alphabet::RFC4648 { padding: false },
+        path.name().as_bytes(),
+    )
+    .to_lowercase();
+    Ok(NotebookMeta {
+        path: path.clone().into_owned(),
+        notebook_path,
+        generated_name,
+    })
+}
+
+async fn convert_notebooks<'a, 'b: 'a>(
     env: &PyEnv,
-    submission: &mut AqoraSubmissionConfig,
+    paths: impl IntoIterator<Item = &'a mut PathStr<'b>>,
 ) -> Result<(), NotebookToPythonFunctionError> {
-    for function in submission.refs.values_mut() {
-        convert_notebook(env, function)?;
+    let paths = paths
+        .into_iter()
+        .map(|path| get_meta(env, path).map(|meta| (path, meta)))
+        .collect::<Result<Vec<_>, _>>()?;
+
+    let mut to_convert = paths.iter().collect::<Vec<_>>();
+    to_convert.dedup_by_key(|(path, _)| path);
+
+    let converted =
+        futures::future::try_join_all(to_convert.into_iter().map(|(_, meta)| async move {
+            notebook_to_script(env, &meta.notebook_path, meta.converted_path()?).await?;
+            let script_path = meta.script_path()?;
+            tokio::fs::write(&script_path, meta.script_function_def())
+                .await
+                .map_err(|e| NotebookToPythonFunctionError::Read(script_path.clone(), e))?;
+            Result::<_, NotebookToPythonFunctionError>::Ok((
```

Comment:
Maybe this could be hoisted next to `let converted` to improve readability? This won't block my approval though

Reply:
you mean the result type? i don't think it can, because you have to know the async function result type to evaluate the `?`. I could put it next to the map function with something like

```rust
.map::<impl Future<Output = Result<_, NotebookToPythonFunctionError>>>(...
```

but i don't know if that helps much, but maybe i'm not understanding your question.